src/ruby.ts

```diff
@@ -0,0 +1,83 @@
+import { exec } from "child_process";
+import { promisify } from "util";
+import * as fs from "fs";
+
+import * as vscode from "vscode";
+
+const asyncExec = promisify(exec);
+const asyncReadFile = promisify(fs.readFile);
+
+export class Ruby {
+  private workingFolder: string;
+  private managerConfiguration: { [key: string]: string };
+
+  constructor() {
+    this.workingFolder = vscode.workspace.workspaceFolders![0].uri.fsPath;
+
+    this.managerConfiguration = vscode.workspace
+      .getConfiguration("rubyLsp")
+      .get("rubyVersionManager")!;
+  }
+
+  async activateRuby() {
+    switch (this.managerConfiguration.manager) {
+      case "chruby":
+        await this.activateChruby();
+        break;
+      default:
+        this.activateShadowenv();
+        break;
+    }
+  }
+
+  activateShadowenv() {
+    vscode.extensions.getExtension("shopify.vscode-shadowenv")?.activate();
+  }
+
+  async activateChruby() {
+    if (!fs.existsSync(`${this.workingFolder}/.ruby-version`)) {
+      vscode.window.showErrorMessage(
+        "Attempted to activate chruby environment, but no .ruby-version file was found."
+      );
+      return;
+    }
+
+    if (!fs.existsSync(this.managerConfiguration.path)) {
+      vscode.window.showErrorMessage(
+        `Attempted to activate chruby environment, but the path provided does not exist (${this.managerConfiguration.path}).`
+      );
+      return;
+    }
+
+    const rubyVersion = await asyncReadFile(
+      `${this.workingFolder}/.ruby-version`
+    );
+
+    try {
+      // Source the chruby file and print all of the generated environment variables
+      const result = await asyncExec(
+        `source ${this.managerConfiguration.path} &&
+        chruby "${rubyVersion.toString().trim()}" &&
+        echo "RUBY_ROOT=$RUBY_ROOT" &&
+        echo "RUBY_ENGINE=$RUBY_ENGINE" &&
+        echo "RUBYOPT=$RUBYOPT" &&
+        echo "PATH=$PATH" &&
+        echo "GEM_HOME=$GEM_HOME" &&
+        echo "GEM_ROOT=$GEM_ROOT" &&
+        echo "GEM_PATH=$GEM_PATH" &&
+        echo "RUBY_VERSION=$RUBY_VERSION"`
+      );
```

Comment:
Would it make more sense to somehow diff the environment before and after running `chruby` so that we can do all the same changes on `process.env`, instead of keeping a list of env variables we know will change?

I am thinking that we can do a `process.env` before calling this script, and we can do a `ruby -rjson -e "puts JSON.dump(ENV.to_h)"` inside the script (since Ruby should be properly setup now) and JSON parse the output of that in the JS script, so that we can compare the two envs and apply changes.

That could also be a reusable thing to apply around anything like this.

Reply:
Hmm, it seems we don't even need to diff anything, we can just inherit the environment from the script:
```ts
const result = await asyncExec(`.....; ruby -rjson -e "puts JSON.dump(ENV.to_h)"`);
process.env = JSON.parse(result.stdout);
```